### PR TITLE
Set-theoretic tuple types

### PR DESCRIPTION
**Pull request by gldubc**
Adds to the Descr tuple types `{:ok, binary(), integer()}` and open tuple types `{atom(), boolean(), ...}` which represent every tuple of at least two elements that are an atom and a boolean.
 
Tuples are encoded as map records, where the keys are the indices. E.g., type `{integer(), atom()}` is encoded as type `%{0 => integer(), 1 => atom()}` and `{atom(), boolean(), ...}` is encoded as the open map `%{..., 0 => atom(), 1 => boolean()}`.

Emptiness checking is slightly modified to account for the fact that tuple types do not admit explicit optional indices.

### CODE REVIEW

File: lib/elixir/lib/module/types/descr.ex [hunk 1]
```diff
@@ -954,6 +981,197 @@ defmodule Module.Types.Descr do
...
+              empty?(diff) or tuple_empty?(tag, Map.put(fields, index, diff), negs)
+            end
+        end
+      end
+
+      true
+    catch
+      :discard_negative -> tuple_empty?(tag, fields, negs)
+    end
+  end
```
Comment (josevalim) [inline lib/elixir/lib/module/types/descr.ex:1173]: Can we rewrite it to a non-raising versaion like the map one is currently implemented? :)

File: lib/elixir/lib/module/types/descr.ex [hunk 2]
```diff
@@ -672,6 +686,19 @@ defmodule Module.Types.Descr do
     end
   end
 
+  defp remove_optional(type) do
```
Comment (josevalim) [inline lib/elixir/lib/module/types/descr.ex:689]: Let's use the existing pop_optional instead!

File: lib/elixir/lib/module/types/descr.ex [hunk 3]
```diff
@@ -954,6 +981,197 @@ defmodule Module.Types.Descr do
...
+
+    map_new(tag, :maps.from_list(pairs))
+  end
+
+  defp tuple_intersection(left, right), do: map_intersection(left, right)
+  defp tuple_difference(left, right), do: map_difference(left, right)
+  defp tuple_union(left, right), do: map_union(left, right)
+
+  # Same as map_fetch, only the tuple descr field is accessed
+  def tuple_fetch(%{} = descr, key) do
```
Comment (josevalim) [inline lib/elixir/lib/module/types/descr.ex:1012]:
```suggestion
  def tuple_fetch(%{} = descr, key) when is_integer(key) do
```

File: lib/elixir/lib/module/types/descr.ex (hunk 1 above)
Comment (sabiwara) [reply lib/elixir/lib/module/types/descr.ex:1173]: @gldubc FYI https://github.com/elixir-lang/elixir/pull/13560

Comment (gldubc) [reply lib/elixir/lib/module/types/descr.ex:1173]: Yes! I see this now :)

File: lib/elixir/lib/module/types/descr.ex [hunk 4]
```diff
@@ -954,6 +981,197 @@ defmodule Module.Types.Descr do
...
+  defp tuple_empty?(_, _, []), do: throw(:not_empty)
+  defp tuple_empty?(_, _, [{:open, fs} | _]) when fs == %{}, do: true
+
+  defp tuple_empty?(tag, fields, [{neg_tag, neg_fields} | negs]) do
+    try do
+      # Indices that exists in the negative tuple, but not in the positive one
+      for {neg_index, neg_type} <- neg_fields, not is_map_key(fields, neg_index) do
+        cond do
+          # This index is not in the closed positive tuple, so there is no value in common
+          # Tthere are no explicit optional types in tuples
```
Comment (ericmj) [inline lib/elixir/lib/module/types/descr.ex:1132]:
```suggestion
          # There are no explicit optional types in tuples
```

**Comment by josevalim** [comment]
Closing in favor of #13752.

**Pull request closed without merging**